src/server/replica.cc

```diff
@@ -517,43 +538,88 @@ error_code Replica::ConsumeRedisStream() {
 }
 
 error_code Replica::ConsumeDflyStream() {
-  // Request master to transition to stable sync.
-  {
-    ReqSerializer serializer{sock_.get()};
-    serializer.SendCommand(StrCat("DFLY STARTSTABLE ", master_context_.dfly_session_id));
-    RETURN_ON_ERR(serializer.ec());
+  DCHECK(cntx_);  // Context from full sync.
+
+  // Send DFLY STARTSTABLE.
+  if (auto ec = SendNextPhaseRequest(); ec) {
+    cntx_->Error(ec);
   }
 
   // Wait for all flows to finish full sync.
-  for (auto& sub_repl : shard_flows_)
-    sub_repl->sync_fb_.join();
+  JoinAllFlows();
+
+  // Transfer cleanup responsibility to stable state task.
+  {
+    lock_guard lk(cntx_mu_);
+
+    // We were cancelled during transition, fast fail.
+    if (cntx_->IsCancelled()) {
+      return *cntx_;
+    }
+
+    cntx_.emplace([this](const auto& ge) {
+      CloseAllSockets();
+      return true;
+    });
+  }
```

Comment:
Same thing here, If we're getting cancelled, then this might happen when we transfer responsibility for resources

Reply:
The whole process of switching callbacks is confusing as hell as well:

Why can't cancellation happen while we hold the lock? 
Because we already joined all the flow fibers, and no error prone operations are currently executed. If we are cancelled externally via Replica::Stop, then it will block on the mutex. This allows us to safely exchange the handlers.

Why not implement a general handler that stores all its state in the replica? This is a feasible solution as well, _however_, the intialization of resources needs to be guarded as well, because, again, in case of external cancellation during initialization we can catch a race condition. 

What comes to mind, is that we can get rid of locking, if we always issue external cancellation on the coordinator fiber (with `socket_->proactoor->Await()`)...

Reply:
> What comes to mind, is that we can get rid of locking, if we always issue external cancellation on the replica coordinator thread (with socket_->proactoor->Await())...

This is probably the way to go

Reply:
why not add those comments to the code? (the explanation suggestion part)

Reply:
Because I didn't yet decide how to implement it, having trouble with the second configuration